Fix routePrefix handling for routes using regex

felipesabino opened the pull request:
Routes using regex had some issues:

- They didn't work when `routePrefix` was provided starting with `/`
- The did work if `/` was omitted from `routePrefix`, but then all other routes started failing

The following changes were made to target this issue:

- Changed the way the full route path was created at `ActionMetadata` to match the resulting regex pattern with those routes created without regex usage.
- Added missing tests for regex routes for an an app that sets the `routePrefix` param

pleerock commented:
@19majkel94 are you able to review this PR?

19majkel94 commented:
@pleerock yeah, why not? I only know some parts of routing-controllers but this PR diff is quite simple 😉

19majkel94 commented on test/functional/global-options.spec.ts:
```diff
@@ -84,13 +89,18 @@ describe("routing-controllers global options", () => {
     
-        assertRequest([3001, 3002], "post", "api/users", { firstName: "Umed", lastName: "Khudoiberdiev" }, response => {
+        assertRequest([3001, 3002, 3003, 3004], "post", "api/users", { firstName: "Umed", lastName: "Khudoiberdiev" }, response => {
+            expect(response).to.have.status(200);
```
You should add `expect(initializedUser).to.be.instanceOf(User)` or sth like that to ensure that the 200 is not a coincidence.

19majkel94 replied:
For both tests! I know that it wasn't in original tests but you have `initializedUser = user;` so make use of it or return custom response string and assert it in test.

felipesabino replied:
Sure! will do

19majkel94 replied:
@felipesabino any progres? I would like to approve this PR but you have to fix the tests 😉

19majkel94 commented on src/metadata/ActionMetadata.ts:
```diff
@@ -256,7 +256,7 @@ export class ActionMetadata {
     static appendBaseRouteToRegexpRoute(route: RegExp, baseRoute: string) {
         if (!baseRoute || baseRoute === "") return route;
-        const fullPath = baseRoute.replace("\/", "\\\\/") + route.toString().substr(1);
+        const fullPath = `^${baseRoute}${route.toString().substr(1)}?$`;
```
Could you describe more this stuff? Why you had to change it, maybe sample `route` and `baseRoute` values please.

felipesabino replied:
@19majkel94 Do you mean as comments on code itself for future references or just here on the PR?

I am asking because I'd rather it on the code, I got very confused with the original `.replace("\/", "\\\\/")` that seemed to be not doing anything

felipesabino replied:
Anyway, this is basically what [expressjs does under the hood](https://github.com/expressjs/express/blob/master/lib/router/layer.js#L16) but using [`path-to-regexp`](https://github.com/pillarjs/path-to-regexp) and the paths where not working properly due to a mismatch on the way the regex was being handled before. 

As `routing-controllers` is handling the route paths as strings directly, do you think it is better to add `path-to-regexp` as a dependency and use it instead of handling the regex manually?

19majkel94 replied:
I don't think so, @pleerock doesn't like dependecies so it won't pass 😜 

If you claim that `.replace("\/", "\\\\/")` was useless it's ok now.
Sample `baseRoute` is `/api` I think, but why `substr(1)` from regex?

pleerock replied:
I don't remember why `.replace("\/", "\\\\/")` is there, maybe its 	
excessive. (tip of myself: comment strange peace of codes 😆 )

pleerock commented:
ping @felipesabino

pleerock commented:
ahh @19majkel94 what shall we do with this PR? Looks like author is missing. Shall we merge it, what do you think?

19majkel94 commented:
I will fix the tests (or you 😜, I don't have write access to the PR branch) for better assertion as I've suggested [here](https://github.com/pleerock/routing-controllers/pull/98#discussion_r107241460) and it will be ready to merge.

pleerock commented:
> or you 😜

simply create fork of his fork and create a new PR 😜😜😜

19majkel94 commented:
> simply create fork of his fork

![](http://i1.kym-cdn.com/photos/images/facebook/000/531/557/a88.jpg)

It's just one line assertion to add, don't complicate things with another forks, branches and PRs 😉

19majkel94 commented:
Ok, If we don't know what `.replace("\/", "\\\\/")` did, we can now resolve the conflicts and merge this PR @pleerock 😉 Thanks @felipesabino 🎉

felipesabino commented:
Sorry guys, I had some issue that kept away from public/open source repos the latest weeks and now it seems that not targeting your comment in a timely fashion screwed up my PR.

I fixed the tests noticed that the way routes are being handled changed significantly, but it is much better than before, congrats! Now I only need to add my changes on `ActionMetadata`, without having to duplicate it for each driver 🎉 

Anyway, I updated everything but could not figure it out why [`dynamic redirect` tests are failing for node latest](https://travis-ci.org/pleerock/routing-controllers/jobs/238436233#L1320) as I could not replicate it locally yet, so I am working on it now. If anybody can help me figure it out, I will be very grateful

pleerock commented:
adding @Diluka can you please take a look why dynamic redirects are failing here?

Diluka commented:
@felipesabino 
Maybe there is something wrong with your network. I edit it to use local service instead of access github api.

**here is the patch**
```patch
Index: test/functional/redirect-decorator.spec.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- test/functional/redirect-decorator.spec.ts	(revision 857e1d158c36c07b0a284c419514bb341b36028f)
+++ test/functional/redirect-decorator.spec.ts	(revision )
@@ -4,6 +4,7 @@
 import {assertRequest} from "./test-utils";
 import {Redirect} from "../../src/decorator/Redirect";
 import {JsonController} from "../../src/decorator/JsonController";
+import {Param} from "../../src/decorator/Param";
 const chakram = require("chakram");
 const expect = chakram.expect;
 
@@ -14,18 +15,31 @@
         // reset metadata args storage
         getMetadataArgsStorage().reset();
 
+        @JsonController("/users")
+        class TestController {
+
+            @Get("/:id")
+            async getOne(@Param("id") id: string) {
+                return {
+                    login: id
+                };
+            }
+
+
+        }
+
         @JsonController()
         class RedirectController {
 
             @Get("/template")
-            @Redirect("https://api.github.com/users/:owner")
+            @Redirect("/users/:owner")
             template() {
                 // console.log("/template");
                 return {owner: "pleerock", repo: "routing-controllers"};
             }
 
             @Get("/original")
-            @Redirect("https://api.github.com/users/pleerock")
+            @Redirect("/users/pleerock")
             original() {
                 // console.log("/original");
             }
@@ -34,7 +48,7 @@
             @Redirect("https://api.github.com")
             override() {
                 // console.log("/override");
-                return "https://api.github.com/users/pleerock";
+                return "/users/pleerock";
             }
 
         }
@@ -54,36 +68,22 @@
     });
     after(done => koaApp.close(done));
 
-    this.timeout(5000); // 2000ms timeout
-
     describe("using template", () => {
-        assertRequest([3001, 3002], "get", "template", {
-            headers: {
-                "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36"
-            }
-        }, response => {
+        assertRequest([3001, 3002], "get", "template", response => {
             expect(response).to.have.status(200);
             expect(response.body).has.property("login", "pleerock");
         });
     });
 
     describe("using override", () => {
-        assertRequest([3001, 3002], "get", "override", {
-            headers: {
-                "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36"
-            }
-        }, response => {
+        assertRequest([3001, 3002], "get", "override", response => {
             expect(response).to.have.status(200);
             expect(response.body).has.property("login", "pleerock");
         });
     });
 
     describe("using original", () => {
-        assertRequest([3001, 3002], "get", "original", {
-            headers: {
-                "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36"
-            }
-        }, response => {
+        assertRequest([3001, 3002], "get", "original", response => {
             expect(response).to.have.status(200);
             expect(response.body).has.property("login", "pleerock");
         });

```

felipesabino commented:
Hey @Diluka,

Thanks for taking a look at it! I don't know if it matters but my network was fine, the problem happened only on CI's environment. Anyways, I applied the patch and it seems to have worked. It doesn't hurt also that it is much faster now 👌 

So @pleerock, I guess it can be merged now if everything looks good to you. Thanks a lot for your patience :1st_place_medal:

pleerock commented:
@19majkel94 can you give me a 🚦 to merge this PR?

19majkel94 commented:
> we can now resolve the conflicts and merge this PR @pleerock 

So "All checks have passed" -> we can merge 💚
